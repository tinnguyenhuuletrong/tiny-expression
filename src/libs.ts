export enum ENodeType {
  Evariable,
  EConst,
  EUnaryOp,
  EUFuncOp,
  EBinaryOp,
  EArrayOp,
}
type OpType =
  | ENodeType.EUnaryOp
  | ENodeType.EUFuncOp
  | ENodeType.EBinaryOp
  | ENodeType.EArrayOp;

export enum EInstructionType {
  plus = "plus",
  minus = "minus",
  mul = "mul",
  div = "div",
  mod = "mod",
  power = "power",

  and = "and",
  or = "or",

  lessorequal = "lessorequal",
  greaterorequal = "greaterorequal",
  equal = "equal",
  notequal = "notequal",
  less = "less",
  greater = "greater",
  negate = "negate",

  contains = "contains",
  notcontains = "notcontains",
  anyof = "anyof",
  allof = "allof",
  empty = "empty",
  notempty = "notempty",
}
export type EUnaryInstructionType =
  | EInstructionType.negate
  | EInstructionType.empty
  | EInstructionType.notempty;

const ALL_UNARY_INS = [
  EInstructionType.negate,
  EInstructionType.empty,
  EInstructionType.notempty,
];

export type TargetValue = Variable | Const | Operand;

export class Variable {
  public readonly type: ENodeType = ENodeType.Evariable;
  constructor(public readonly name: string) {}

  toObject() {
    const { type, name } = this;
    return { type, name };
  }
  static fromObj(obj: any) {
    return new Variable(obj.name);
  }
}

export class Const {
  public readonly type: ENodeType = ENodeType.EConst;
  constructor(public readonly value: string | number | boolean) {}

  toObject() {
    const { type, value } = this;
    return { type, value };
  }
  static fromObj(obj: any) {
    return new Const(obj.value);
  }
}

export abstract class Operand {
  public readonly type!: OpType;
  constructor(type: OpType) {
    this.type = type;
  }
  toObject(): any {
    return { type: this.type };
  }
}

export class FunctionOperand extends Operand {
  constructor(
    public readonly name: string,
    public readonly params: ArrayOperand
  ) {
    super(ENodeType.EUFuncOp);
  }

  toObject() {
    const { name, params } = this;
    return { ...super.toObject(), name, params: params.toObject() };
  }
  static fromObj(obj: any) {
    const params = valueFactory(obj.params);
    _ensureType(params, ENodeType.EArrayOp);
    return new FunctionOperand(obj.name, params as ArrayOperand);
  }
}

export class UnaryOperand extends Operand {
  public readonly op: EUnaryInstructionType;
  public readonly target: TargetValue;

  constructor(target: TargetValue | null, op: EInstructionType) {
    super(ENodeType.EUnaryOp);

    if (!target) throw new Error(`UnaryOperand target is null`);
    if (!ALL_UNARY_INS.includes(op))
      throw new Error(`UnaryOperand not accept insType ${op}`);

    this.target = target;
    this.op = op as any;
  }

  toObject() {
    const { op, target } = this;
    return { ...super.toObject(), op, target: target.toObject() };
  }
  static fromObj(obj: any) {
    const target = valueFactory(obj.target);
    return new UnaryOperand(target, obj.op);
  }
}

export class BinaryOperand extends Operand {
  public readonly op: EInstructionType;
  public readonly left: TargetValue;
  public readonly right: TargetValue;
  public readonly isArithmeticOp: boolean;
  constructor(
    op: EInstructionType,
    left: TargetValue,
    right: TargetValue,
    isArithmeticOp: boolean = false
  ) {
    super(ENodeType.EBinaryOp);
    this.op = op;
    this.left = left;
    this.right = right;
    this.isArithmeticOp = isArithmeticOp;
  }

  toObject() {
    const { op, left, right, isArithmeticOp } = this;
    return {
      ...super.toObject(),
      op,
      left: left.toObject(),
      right: right.toObject(),
      isArithmeticOp,
    };
  }
  static fromObj(obj: any) {
    const op = obj.op;
    const left = valueFactory(obj.left);
    const right = valueFactory(obj.right);
    const isArithmeticOp = Boolean(obj.isArithmeticOp);
    return new BinaryOperand(op, left, right, isArithmeticOp);
  }
}

export class ArrayOperand extends Operand {
  public readonly data: TargetValue[];
  constructor(params: TargetValue[]) {
    super(ENodeType.EArrayOp);
    this.data = params;
  }

  toObject() {
    const { data } = this;
    return {
      ...super.toObject(),
      data: data.map((itm) => itm.toObject()),
    };
  }
  static fromObj(obj: any) {
    if (!Array.isArray(obj.data))
      throw new Error("ArrayOperand parse error. obj.data must be array");
    const data = obj.data.map((itm: any) => valueFactory(itm));
    return new ArrayOperand(data);
  }
}

export function buildBinaryOperand(
  head: any,
  tail: Array<any>,
  isArithmeticOp: boolean = false
): TargetValue {
  return tail.reduce(function (result, elements) {
    return new BinaryOperand(
      elements.v1.value,
      result,
      elements.v2.value,
      isArithmeticOp
    );
  }, head.value);
}

export function flattenArray(array: Array<any>): Array<any> {
  return [].concat.apply([], array);
}

function _ensureType(obj: any, t: ENodeType) {
  if (obj?.type !== t) throw new Error(`expected type ${t}`);
}

export function valueFactory(obj: any): TargetValue {
  const type = obj.type as ENodeType;
  switch (type) {
    case ENodeType.EConst:
      return Const.fromObj(obj);
    case ENodeType.Evariable:
      return Variable.fromObj(obj);
    case ENodeType.EUnaryOp:
      return UnaryOperand.fromObj(obj);
    case ENodeType.EArrayOp:
      return ArrayOperand.fromObj(obj);
    case ENodeType.EBinaryOp:
      return BinaryOperand.fromObj(obj);
    case ENodeType.EUFuncOp:
      return FunctionOperand.fromObj(obj);
    default:
      // should not reachable
      ((_x: never) => {})(type);
      throw new Error(`unknown type ${type}`);
  }
}
