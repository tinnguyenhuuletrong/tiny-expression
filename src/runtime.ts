import {
  ArrayOperand,
  BinaryOperand,
  Const,
  EInstructionType,
  ENodeType,
  FunctionOperand,
  TargetValue,
  UnaryOperand,
  Variable,
  EUnaryInstructionType,
} from "./libs";
import {
  _castToArray,
  _castToBoolean,
  _castToStringOrNumber,
  _castToNumber,
} from "./utils";

export type RuntimeValue =
  | string
  | number
  | boolean
  | Array<RuntimeValue>
  | null;

export interface IRuntimeContext {
  resolveVariable(name: string): RuntimeValue;
  functionCall(name: string, params: RuntimeValue[]): RuntimeValue;
}

export class Runtime {
  constructor(private ctx: IRuntimeContext) {}

  eval(root: TargetValue) {
    return this._internalEval(root);
  }

  private _internalEval(node: TargetValue): RuntimeValue {
    const type = node.type;
    switch (type) {
      case ENodeType.EConst: {
        const ins = node as Const;
        return ins.value;
      }
      case ENodeType.Evariable: {
        const ins = node as Variable;
        return this.ctx.resolveVariable(ins.name);
      }
      case ENodeType.EBinaryOp: {
        const ins = node as BinaryOperand;
        return this._calBinaryOp(ins);
      }
      case ENodeType.EUnaryOp: {
        const ins = node as UnaryOperand;
        return this._calUnaryOp(ins);
      }
      case ENodeType.EUFuncOp: {
        const ins = node as FunctionOperand;
        const params = _castToArray(this._internalEval(ins.params));
        return this.ctx.functionCall(ins.name, params);
      }
      case ENodeType.EArrayOp: {
        const ins = node as ArrayOperand;
        const tmp = ins.data.map((itm) => this._internalEval(itm));
        return tmp;
      }

      default:
        // should not reachable
        ((_x: never) => {})(type);
        throw new Error(`unknown type ${type}`);
    }
  }

  private _calUnaryOp(ins: UnaryOperand): RuntimeValue {
    const op: EUnaryInstructionType = ins.op;
    switch (op) {
      case EInstructionType.negate: {
        const v = _castToBoolean(this._internalEval(ins.target));
        return !v;
      }
      case EInstructionType.empty: {
        const arr = _castToArray(this._internalEval(ins.target));
        return arr.length === 0;
      }
      case EInstructionType.notempty: {
        const arr = _castToArray(this._internalEval(ins.target));
        return arr.length !== 0;
      }
    }
  }

  private _calBinaryOp(ins: BinaryOperand): RuntimeValue {
    const op = ins.op;
    switch (op) {
      case EInstructionType.plus: {
        const left: any = _castToStringOrNumber(this._internalEval(ins.left));
        const right = _castToStringOrNumber(this._internalEval(ins.right));
        return left + right;
      }
      case EInstructionType.minus: {
        const left = _castToNumber(this._internalEval(ins.left));
        const right = _castToNumber(this._internalEval(ins.right));
        return left - right;
      }
      case EInstructionType.mul: {
        const left = _castToNumber(this._internalEval(ins.left));
        const right = _castToNumber(this._internalEval(ins.right));
        return left * right;
      }
      case EInstructionType.div: {
        const left = _castToNumber(this._internalEval(ins.left));
        const right = _castToNumber(this._internalEval(ins.right));
        return left / right;
      }
      case EInstructionType.mod: {
        const left = _castToNumber(this._internalEval(ins.left));
        const right = _castToNumber(this._internalEval(ins.right));
        return left % right;
      }
      case EInstructionType.power: {
        const left = _castToNumber(this._internalEval(ins.left));
        const right = _castToNumber(this._internalEval(ins.right));
        return left ** right;
      }
      case EInstructionType.equal: {
        const left = this._internalEval(ins.left);
        const right = this._internalEval(ins.right);
        return left == right;
      }
      case EInstructionType.notequal: {
        const left = this._internalEval(ins.left);
        const right = this._internalEval(ins.right);
        return left != right;
      }
      case EInstructionType.greater: {
        const left = _castToStringOrNumber(this._internalEval(ins.left));
        const right = _castToStringOrNumber(this._internalEval(ins.right));
        return left > right;
      }
      case EInstructionType.greaterorequal: {
        const left = _castToStringOrNumber(this._internalEval(ins.left));
        const right = _castToStringOrNumber(this._internalEval(ins.right));
        return left >= right;
      }
      case EInstructionType.less: {
        const left = _castToStringOrNumber(this._internalEval(ins.left));
        const right = _castToStringOrNumber(this._internalEval(ins.right));
        return left < right;
      }
      case EInstructionType.lessorequal: {
        const left = _castToStringOrNumber(this._internalEval(ins.left));
        const right = _castToStringOrNumber(this._internalEval(ins.right));
        return left <= right;
      }
      case EInstructionType.and: {
        return (
          _castToBoolean(this._internalEval(ins.left)) &&
          _castToBoolean(this._internalEval(ins.right))
        );
      }
      case EInstructionType.or: {
        return (
          _castToBoolean(this._internalEval(ins.left)) ||
          _castToBoolean(this._internalEval(ins.right))
        );
      }

      // isArithmeticOp == false
      case EInstructionType.contains: {
        const arr = _castToArray(this._internalEval(ins.left));
        const val = this._internalEval(ins.right);
        return arr.includes(val);
      }
      case EInstructionType.notcontains: {
        const arr = _castToArray(this._internalEval(ins.left));
        const val = this._internalEval(ins.right);
        return !arr.includes(val);
      }
      case EInstructionType.allof: {
        const arr = _castToArray(this._internalEval(ins.left));
        const val = this._internalEval(ins.right);
        for (const it of arr) {
          if (it !== val) return false;
        }
        return true;
      }
      case EInstructionType.anyof: {
        const arr = _castToArray(this._internalEval(ins.left));
        const val = this._internalEval(ins.right);
        for (const it of arr) {
          if (it === val) return true;
        }
        return false;
      }

      default:
        return null;
    }
  }
}
