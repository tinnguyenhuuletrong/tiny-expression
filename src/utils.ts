import type { RuntimeValue } from ".";
import { TargetValue, valueFactory } from "./libs";

export function serializeToJson(root: TargetValue): string {
  return JSON.stringify(root.toObject());
}

export function deserializeFromJson(str: string): TargetValue | null {
  const obj = JSON.parse(str);
  if (obj === null || obj === undefined) return null;
  return valueFactory(obj);
}

export function _castToStringOrNumber(v: RuntimeValue): string | number {
  const tmp = typeof v;
  if (tmp === "number" || tmp === "string") return v as string | number;
  throw new Error(`${v} is not TypePrimitive. it is ${tmp}`);
}

export function _castToNumber(v: RuntimeValue): number {
  const tmp = typeof v;
  if (tmp === "number") return v as number;
  throw new Error(`${v} is not TypePrimitive`);
}

export function _castToBoolean(v: RuntimeValue): boolean {
  return Boolean(v);
}

export function _castToDate(v: RuntimeValue): Date {
  let d!: Date;
  if (typeof v === "number") {
    d = new Date(v);
  } else if (typeof v === "string") {
    d = new Date(v);
    if (isNaN(d.valueOf())) {
      throw new Error(`${v} - ${d} is not a Date like string`);
    }
  } else {
    throw new Error(`${v} is not numer or string. can not cast to date`);
  }
  return d;
}

export function _castToArray(v: RuntimeValue): Array<RuntimeValue> {
  if (Array.isArray(v)) return v;
  throw new Error("ensureArray invalid cast");
}
