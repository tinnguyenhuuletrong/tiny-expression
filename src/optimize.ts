import {
  ArrayOperand,
  BinaryOperand,
  Const,
  EInstructionType,
  ENodeType,
  FunctionOperand,
  TargetValue,
  UnaryOperand,
} from "./libs";
import { _castToArray, _castToBoolean, _castToStringOrNumber } from "./utils";

export function doOptimize(node: TargetValue): TargetValue {
  return _doOptimizeInternal(node);
}

function _doOptimizeInternal(node: TargetValue): TargetValue {
  const type = node.type;
  switch (type) {
    case ENodeType.EBinaryOp: {
      const tmp = node as BinaryOperand;
      const left = _doOptimizeInternal(tmp.left);
      const right = _doOptimizeInternal(tmp.right);

      const canOptimize = isConst(left) && isConst(right);
      if (!canOptimize) {
        return new BinaryOperand(tmp.op, left, right, tmp.isArithmeticOp);
      }

      switch (tmp.op) {
        case EInstructionType.plus: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 + v2);
        }
        case EInstructionType.minus: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 - v2);
        }
        case EInstructionType.mul: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 * v2);
        }
        case EInstructionType.div: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 / v2);
        }
        case EInstructionType.mod: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 % v2);
        }
        case EInstructionType.power: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 ** v2);
        }

        default:
          return node;
      }
    }
    case ENodeType.EUnaryOp: {
      const tmp = node as UnaryOperand;
      const target = _doOptimizeInternal(tmp.target);

      const canOptimize = isConst(target);
      if (!canOptimize) return node;

      switch (tmp.op) {
        case EInstructionType.negate:
          const v = _castToBoolean((target as Const).value);
          return new Const(!v);

        default:
          return node;
      }
    }
    case ENodeType.EArrayOp: {
      const ins = node as ArrayOperand;
      const newData = ins.data.map((itm) => _doOptimizeInternal(itm));
      return new ArrayOperand(newData);
    }
    case ENodeType.EUFuncOp: {
      const ins = node as FunctionOperand;
      const params = _doOptimizeInternal(ins.params);
      return new FunctionOperand(ins.name, params as ArrayOperand);
    }
    default:
      return node;
  }
}

function isConst(node: TargetValue) {
  return node.type === ENodeType.EConst;
}
