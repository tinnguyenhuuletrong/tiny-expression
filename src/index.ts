import { parse, Parser } from "./_tinyExpressionGrammar";
import { doOptimize } from "./optimize";
export { Runtime, type RuntimeValue, type IRuntimeContext } from "./runtime";
export { deserializeFromJson, serializeToJson } from "./utils";
export * from "./addons/StandardRuntimeContext";
export { parse, Parser, doOptimize };

export function compileAndOptimize(inp: string) {
  const tmp = compile(inp);
  try {
    const optimized = doOptimize(tmp);
    return optimized;
  } catch (error) {
    throw new Error(`optimized error ${inp} -> ${String(error)}`);
  }
}

export function compile(inp: string) {
  const tmp = parse(inp);
  if (!tmp.ast) {
    const msg = tmp.errs.map((itm) => itm.toString()).join(",");
    throw new Error(`compile error ${inp} -> ${msg}`);
  }
  return tmp.ast.value;
}
