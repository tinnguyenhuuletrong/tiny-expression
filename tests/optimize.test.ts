import { test, expect } from "bun:test";
import { IRuntimeContext, RuntimeValue, compileAndOptimize } from "../src";

const ALL_CASES_OPTIMIZE: string[] = [
  `( (1*3 / 2) ^ 7 ) % ( {val} - 3 )`,
  `iff(!true, 10*3, 10*2)`,
  `[!true, !!false, [9*3+7], log((-2 + 2) * {age})]`,
];

test("do optimize", async () => {
  for (const inp of ALL_CASES_OPTIMIZE) {
    const res = compileAndOptimize(inp);
    console.log("\t", inp, "->", res);
    expect(res).toMatchSnapshot();
  }
});
