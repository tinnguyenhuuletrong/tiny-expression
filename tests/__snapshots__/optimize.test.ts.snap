// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`do optimize 1`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 17.0859375,
  },
  "op": "mod",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": Variable {
      "name": "val",
      "type": 0,
    },
    "op": "minus",
    "right": Const {
      "type": 1,
      "value": 3,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`do optimize 2`] = `
FunctionOperand {
  "name": "iff",
  "params": ArrayOperand {
    "data": [
      Const {
        "type": 1,
        "value": false,
      },
      Const {
        "type": 1,
        "value": 30,
      },
      Const {
        "type": 1,
        "value": 20,
      },
    ],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`do optimize 3`] = `
ArrayOperand {
  "data": [
    Const {
      "type": 1,
      "value": false,
    },
    Const {
      "type": 1,
      "value": false,
    },
    ArrayOperand {
      "data": [
        Const {
          "type": 1,
          "value": 34,
        },
      ],
      "type": 5,
    },
    FunctionOperand {
      "name": "log",
      "params": ArrayOperand {
        "data": [
          BinaryOperand {
            "isArithmeticOp": true,
            "left": Const {
              "type": 1,
              "value": 0,
            },
            "op": "mul",
            "right": Variable {
              "name": "age",
              "type": 0,
            },
            "type": 4,
          },
        ],
        "type": 5,
      },
      "type": 3,
    },
  ],
  "type": 5,
}
`;
